File: gospider/audit/audit.go
package audit

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"time"
)

type AuditEvent struct {
	Timestamp time.Time      `json:"timestamp"`
	JobID     string         `json:"job_id,omitempty"`
	StepID    string         `json:"step_id,omitempty"`
	Type      string         `json:"type"`
	Payload   map[string]any `json:"payload,omitempty"`
}

type AuditTrail interface {
	Append(event AuditEvent) error
	Events() []AuditEvent
}

type MemoryAuditTrail struct {
	mu     sync.RWMutex
	events []AuditEvent
}

func (t *MemoryAuditTrail) Append(event AuditEvent) error {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.events = append(t.events, event)
	return nil
}

func (t *MemoryAuditTrail) Events() []AuditEvent {
	t.mu.RLock()
	defer t.mu.RUnlock()
	events := make([]AuditEvent, len(t.events))
	copy(events, t.events)
	return events
}

type FileAuditTrail struct {
	Path string
	mu   sync.Mutex
}

func (t *FileAuditTrail) Append(event AuditEvent) error {
	if err := os.MkdirAll(filepath.Dir(t.Path), 0o755); err != nil {
		return err
	}
	data, err := json.Marshal(event)
	if err != nil {
		return err
	}
	t.mu.Lock()
	defer t.mu.Unlock()
	handle, err := os.OpenFile(t.Path, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
	if err != nil {
		return err
	}
	defer handle.Close()
	if _, err := handle.Write(append(data, '\n')); err != nil {
		return err
	}
	return handle.Sync()
}

func (t *FileAuditTrail) Events() []AuditEvent {
	data, err := os.ReadFile(t.Path)
	if err != nil {
		return nil
	}
	lines := bytesToLines(data)
	events := make([]AuditEvent, 0, len(lines))
	for _, line := range lines {
		var event AuditEvent
		if err := json.Unmarshal(line, &event); err == nil {
			events = append(events, event)
		}
	}
	return events
}

type CompositeAuditTrail struct {
	trails []AuditTrail
}

func NewCompositeAuditTrail(trails ...AuditTrail) *CompositeAuditTrail {
	return &CompositeAuditTrail{trails: trails}
}

func (t *CompositeAuditTrail) Append(event AuditEvent) error {
	for _, trail := range t.trails {
		if trail == nil {
			continue
		}
		if err := trail.Append(event); err != nil {
			return err
		}
	}
	return nil
}

func (t *CompositeAuditTrail) Events() []AuditEvent {
	var all []AuditEvent
	for _, trail := range t.trails {
		if trail == nil {
			continue
		}
		all = append(all, trail.Events()...)
	}
	return all
}

func bytesToLines(data []byte) [][]byte {
	lines := make([][]byte, 0)
	start := 0
	for index, b := range data {
		if b == '\n' {
			if index > start {
				lines = append(lines, data[start:index])
			}
			start = index + 1
		}
	}
	if start < len(data) {
		lines = append(lines, data[start:])
	}
	return lines
}

File: pyspider/runtime/audit.py
from __future__ import annotations

import json
from dataclasses import asdict, dataclass, is_dataclass
from pathlib import Path
from threading import RLock
from typing import Any, Iterable, List, Protocol


@dataclass(slots=True)
class AuditEvent:
    timestamp: str
    job_id: str = ""
    step_id: str = ""
    event_type: str = ""
    payload: dict[str, Any] | None = None


class AuditTrail(Protocol):
    def append(self, event: AuditEvent) -> None: ...
    def events(self) -> List[AuditEvent]: ...


class MemoryAuditTrail:
    def __init__(self) -> None:
        self._lock = RLock()
        self._events: List[AuditEvent] = []

    def append(self, event: AuditEvent) -> None:
        with self._lock:
            self._events.append(event)

    def events(self) -> List[AuditEvent]:
        with self._lock:
            return list(self._events)


class FileAuditTrail:
    def __init__(self, path: str | Path) -> None:
        self.path = Path(path)
        self._lock = RLock()

    def append(self, event: AuditEvent) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        with self._lock, self.path.open("a", encoding="utf-8") as handle:
            handle.write(json.dumps(_normalize(event), ensure_ascii=False) + "\n")

    def events(self) -> List[AuditEvent]:
        if not self.path.exists():
            return []
        items = []
        with self._lock:
            for line in self.path.read_text(encoding="utf-8").splitlines():
                if not line.strip():
                    continue
                payload = json.loads(line)
                items.append(
                    AuditEvent(
                        timestamp=payload.get("timestamp", ""),
                        job_id=payload.get("job_id", ""),
                        step_id=payload.get("step_id", ""),
                        event_type=payload.get("event_type", ""),
                        payload=payload.get("payload"),
                    )
                )
        return items


class CompositeAuditTrail:
    def __init__(self, trails: Iterable[AuditTrail]) -> None:
        self.trails = list(trails)

    def append(self, event: AuditEvent) -> None:
        for trail in self.trails:
            trail.append(event)

    def events(self) -> List[AuditEvent]:
        events: List[AuditEvent] = []
        for trail in self.trails:
            events.extend(trail.events())
        return events


def _normalize(value: Any) -> Any:
    if is_dataclass(value):
        return asdict(value)
    if isinstance(value, dict):
        return {str(key): _normalize(item) for key, item in value.items()}
    if isinstance(value, (list, tuple)):
        return [_normalize(item) for item in value]
    return value
